### PR TITLE
issue #17259 - move rcptbycustgrp to core

### PR DESCRIPTION
**Pull request by bartmaj**
Transfer of Receipt by Customer Group extension to the core application.

Database changes at xtuple/xtuple#2361

### CODE REVIEW

**Comment by bartmaj** [comment]
xtuple/xtuple#2361

**Comment by andersdd** [comment]
Check with @mccjim whether the Credit Memo item and Invoice Item modifications should be ported into core.  They were *not* an integral part of the Cash Receipt by Customer group modification but were included at the same time/package as it was a requirement by the same customer.

BTW I was the original author of this package.

File: guiclient/cashReceipt.cpp [hunk 1]
```diff
@@ -157,6 +171,216 @@ cashReceipt::cashReceipt(QWidget* parent, const char* name, Qt::WindowFlags fl)
   _posted = false;
 }
 
+
+void cashReceipt::activateButtons(bool c = true)
```
Comment (gilmoskowitz) [inline guiclient/cashReceipt.cpp]: Supplying a default value for `c` should go in `cashreceipt.h`, not here.

File: guiclient/cashReceipt.cpp [hunk 2]
```diff
@@ -157,6 +171,216 @@ cashReceipt::cashReceipt(QWidget* parent, const char* name, Qt::WindowFlags fl)
...
+      sql = "SELECT cust_curr_id FROM custinfo JOIN custgrpitem on cust_id=custgrpitem_cust_id "
+            "WHERE custgrpitem_custgrp_id =<? value ('custgrp_id') ?> LIMIT 1;";
+
+    MetaSQLQuery mql(sql);
+    cust = mql.toQuery(pList);
+    if (cust.first())
+      _received->setId(cust.value("cust_curr_id").toInt());
+    else if (cust.lastError().type() == QSqlError::NoError)
+    {
+      systemError(this, cust.lastError().databaseText(), __FILE__, __LINE__);
```
Comment (gilmoskowitz) [inline guiclient/cashReceipt.cpp]: Please use `ErrorReporter::error()` here and elsewhere for detecting and reporting query errors.

File: guiclient/cashReceipt.cpp [hunk 3]
```diff
@@ -157,6 +171,216 @@ cashReceipt::cashReceipt(QWidget* parent, const char* name, Qt::WindowFlags fl)
...
+}
+
+
+void cashReceipt::updateCustomerGroup()
+{
+  if (_mode == 2 || _mode == cNew) //originally it's supposed to work only for cNew; have to look into it
+  {
+
+    QString sql="UPDATE cashrcpt SET cashrcpt_custgrp_id=<? value (\"custgrp_id\") ?> "
+                "WHERE cashrcpt_number = <? value(\"number\") ?>;";
```
Comment (gilmoskowitz) [inline guiclient/cashReceipt.cpp]: For clarity use single quotes here instead of backslashed double quotes.

File: guiclient/cashReceipt.cpp [hunk 4]
```diff
@@ -157,6 +171,216 @@ cashReceipt::cashReceipt(QWidget* parent, const char* name, Qt::WindowFlags fl)
...
+  bool valid;
+
+  param = pParams.value("mode", &valid);
+  if (valid)
+  {
+    _mode = param.toInt();
+
+    if (param.toString() == "edit" || param.toString() == "view")
+    {
+      _cashrcptid = pParams.value("cashrcpt_id", &valid).toInt();
```
Comment (gilmoskowitz) [inline guiclient/cashReceipt.cpp]: This might crash if `cashrcpt_id` isn't in pParams. Better to split this into two lines and add an early return if `! valid`.

File: guiclient/cashReceipt.cpp [hunk 5]
```diff
@@ -301,21 +530,27 @@ void cashReceipt::sApplyToBalance()
...
   if (applyToBal.lastError().type() != QSqlError::NoError)
       systemError(this, applyToBal.lastError().databaseText(), __FILE__, __LINE__);
 
+  if (_custSelector->code() == "G")
+    updateCustomerGroup();
+
   sFillApplyList();
+
+  if (_custSelector->code() == "G")
+      omfgThis->sCashReceiptsUpdated(1, true);
```
Comment (gilmoskowitz) [inline guiclient/cashReceipt.cpp]: Why is the `sCashReceiptsUpdated()` call specific to customer groups? Either it should always be done or this line should be commented.

File: guiclient/cashReceipt.cpp [hunk 6]
```diff
@@ -462,21 +704,34 @@ void cashReceipt::sClear()
 
   sFillApplyList();
   sUpdateBalance();
+
+  if (_custSelector->code() == "G")
+  {
+      updateCustomerGroup();
+      sFillApplyList();
```
Comment (gilmoskowitz) [inline guiclient/cashReceipt.cpp]: This `sFillApplyList()` looks redundant. Can it go away if the `updateCustomerGroup()` is moved up a few lines?

File: guiclient/cashReceipt.cpp [hunk 7]
```diff
@@ -552,6 +808,15 @@ void cashReceipt::close()
 
 void cashReceipt::sSave()
 {
+  if (_received->localValue() == 0.00 &&
+      QMessageBox::question(this, tr("Invalid amout"),
```
Comment (gilmoskowitz) [inline guiclient/cashReceipt.cpp]: Typo - `amout`

File: guiclient/cashReceiptMiscDistrib.cpp [hunk 8]
```diff
@@ -147,5 +156,83 @@ void cashReceiptMiscDistrib::sSave()
...
+
+  if (_mode == cEdit)
+    _custSelector->setId(customer);
+}
+
+void cashReceiptMiscDistrib::setGroup(const ParameterList &pParams)
+{
+  if (_mode == cNew)
+  {
+    QString sql = "SELECT cashrcpt_custgrp_id FROM cashrcpt WHERE cashrcpt_id=<? value(\"cashrcpt_id\") ?>";
```
Comment (gilmoskowitz) [inline guiclient/cashReceiptMiscDistrib.cpp]: `\"` -> `'`

File: guiclient/cashReceiptMiscDistrib.cpp [hunk 9]
```diff
@@ -147,5 +156,83 @@ void cashReceiptMiscDistrib::sSave()
...
+
+  //Set Up Customer widget
+  XSqlQuery query;
+  query.prepare("SELECT custinfo.cust_id,custinfo.cust_name FROM custinfo, custgrpitem "
+                "WHERE custgrpitem.custgrpitem_cust_id = custinfo.cust_id "
+                "AND (custgrpitem_custgrp_id =:group);");
+  query.bindValue(":group", group);
+  query.exec();
+  if (query.first())
+    _custSelector->populate(query);
```
Comment (gilmoskowitz) [inline guiclient/cashReceiptMiscDistrib.cpp]: Add query error checking.

File: guiclient/creditMemoItem.cpp [hunk 10]
```diff
@@ -168,6 +170,21 @@ enum SetResponse creditMemoItem::set(const ParameterList &pParams)
...
       connect(_item, SIGNAL(valid(bool)), _listPrices, SLOT(setEnabled(bool)));
+
+      //cash_receipt_by_customer_group
+      QString sql = "SELECT * FROM cmitemaccnt WHERE cmitemaccnt_cmhead_id=<? value('cmhead') ?> "
+                    "AND cmitemaccnt_cmitem_id=<? value('cmitem') ?>;";
+      XSqlQuery query;
+      MetaSQLQuery mql(sql);
+      ParameterList qparams;
+      qparams.append("cmhead", _cmheadid);
+      qparams.append("accnt", _revAccnt->id());
```
Comment (gilmoskowitz) [inline guiclient/creditMemoItem.cpp]: The `accnt` parameter isn't used.

File: guiclient/invoiceItem.cpp [hunk 11]
```diff
@@ -335,7 +335,7 @@ void invoiceItem::populate()
     _lineNumber->setText(invcitem.value("invcitem_linenumber").toString());
 
     // TODO: should this check itemsite_controlmethod == N?
-    _trackqoh = (invcitem.value("invcitem_invcitem_id").toInt() > 0 &&
+    _trackqoh = (invcitem.value("invcitem_item_id").toInt() > 0 &&
```
Comment (gilmoskowitz) [inline guiclient/invoiceItem.cpp]: Is this a known bug???

File: guiclient/postCashReceipts.cpp [hunk 12]
```diff
@@ -53,12 +53,17 @@ void postCashReceipts::sPost()
...
+                 "JOIN cashrcptitem ON cashrcpt_id=cashrcptitem_cashrcpt_id "
+                 "JOIN custinfo ON cust_id=cashrcptitem_cust_id "
+                 "WHERE ((NOT cashrcpt_posted) AND (NOT cashrcpt_void));");
+                 */
   postPost.exec( "SELECT cashrcpt_id, cust_number "
-          "FROM cashrcpt, custinfo "
-          "WHERE ( (NOT cashrcpt_posted)"
-          "  AND   (NOT cashrcpt_void)"
-          "  AND   (cashrcpt_cust_id=cust_id) );" );
+                 "FROM cashrcp, custinfo "
```
Comment (gilmoskowitz) [inline guiclient/postCashReceipts.cpp]: Typo `cashrcp`?

File: guiclient/printStatementsByCustomerGroup.h [hunk 13]
```diff
@@ -0,0 +1,28 @@
+#ifndef PRINTSTATEMENTSBYCUSTOMERGROUP
```
Comment (gilmoskowitz) [inline guiclient/printStatementsByCustomerGroup.h]: Missing standard license/copyright comment.

File: guiclient/cashReceipt.cpp [hunk 14]
```diff
@@ -88,6 +92,11 @@ cashReceipt::cashReceipt(QWidget* parent, const char* name, Qt::WindowFlags fl)
   bg->addButton(_balCreditMemo);
   bg->addButton(_balCustomerDeposit);
 
+  //Set up Customer/Customer Group widget
+  _custSelector->populate("SELECT 1 as selkey, 'Customer' as selname, 'C' as selcode"
+                          " UNION "
+                          "SELECT 2  as selkey, 'Customer Group' as selname, 'G' as selcode;");
+
```
Comment (andersdd) [inline guiclient/cashReceipt.cpp]: @gilmoskowitz this pseudo widget is basically the CustomerSelector widget without the options for selecting by customer type or all customers.  Would it be worth creating an actual widget variant for this?  It would be basically a cut and paste and removal of unneeded elements so quite simple to do.

File: guiclient/cashReceipt.cpp [hunk 15]
```diff
@@ -157,6 +171,216 @@ cashReceipt::cashReceipt(QWidget* parent, const char* name, Qt::WindowFlags fl)
...
+//  Need to find the Cash Receipt id
+  if (_mode == cNew)
+  {
+    XSqlQuery query;
+    QString sql = "SELECT cashrcpt_id FROM cashrcpt WHERE cashrcpt_number=<? value(\"number\") ?>;";
+    MetaSQLQuery mql(sql);
+    query = mql.toQuery(getParams());
+    if (query.first())
+      _cashrcptid = query.value("cashrcpt_id").toInt();
+  }
```
Comment (andersdd) [inline guiclient/cashReceipt.cpp]: The above code was required by scripting as we do not have access to the _cashrcptid variable.  Not needed in core.  Use _cashrcptid instead.

File: guiclient/cashReceipt.cpp [hunk 16]
```diff
@@ -157,6 +171,216 @@ cashReceipt::cashReceipt(QWidget* parent, const char* name, Qt::WindowFlags fl)
...
+  if (_mode == 2 || _mode == cNew) //originally it's supposed to work only for cNew; have to look into it
+  {
+
+    QString sql="UPDATE cashrcpt SET cashrcpt_custgrp_id=<? value (\"custgrp_id\") ?> "
+                "WHERE cashrcpt_number = <? value(\"number\") ?>;";
+    XSqlQuery query;
+    MetaSQLQuery mql(sql);
+    ParameterList p;
+    p=getParams();
+    query = mql.toQuery(getParams());
```
Comment (andersdd) [inline guiclient/cashReceipt.cpp]: ParameterList p is not used and getParams called twice.  Lines 320:321 not required

File: guiclient/cashReceipt.cpp (hunk 6 above)
Comment (andersdd) [inline guiclient/cashReceipt.cpp]: Agree with @gilmoskowitz,  you should be able to reduce redundancy here.  In the scripted screen we had to _override_ the core fillList function whereas you no longer need to do that.

File: guiclient/cashReceiptMiscDistrib.cpp [hunk 17]
```diff
@@ -147,5 +156,83 @@ void cashReceiptMiscDistrib::sSave()
...
+    if (query.first())
+    {
+      _custgrp = query.value("cashrcpt_custgrp_id").toInt();
+      showCustomers(_custgrp, query.value("cashrcptmisc_cust_id").toInt());
+    }
+  }
+}
+
+void cashReceiptMiscDistrib::gSave() //can be moved to sSave()
+{
```
Comment (andersdd) [inline guiclient/cashReceiptMiscDistrib.cpp]: All save functions should be under sSave().  No longer need gSave() which was used in script as an override or enhanced save function

File: guiclient/invoiceItem.cpp (hunk 11 above)
Comment (andersdd) [inline guiclient/invoiceItem.cpp]: Hard to say if it is an identified bug, but it is the correct fix.

File: guiclient/cashReceipt.cpp (hunk 14 above)
Comment (gilmoskowitz) [inline guiclient/cashReceipt.cpp]: @andersdd - good spotting.

@bartmaj - think about what would be required to expand the API of the existing CustomerSelector to allow code and Qt Designer to choose which subset of options (Customer, Customer Group, Customer Type) should be available to the user. this is recorded as issue #26038, which should be dealt with in a separate pull request.
